prompt:
I have the following count table for an RNA-seq experiment, rows are genes and columns are samples, "Supplementary_Table_1_exp.xlsx". The first column of the file is my gene names and the rest of the columns are samples. My annotation data is here  "SRP021193.colData.tsv", it contains sample annotations and rows of this matches the columns of count table, first column is the sample ids. Merge annotation data and gene expression tables. Plot boxplots for gene expression values per sample and color code boxplots based on group" variable in my annotation data.

here are first few lines of the file(s).

Supplementary_Table_1_exp.xlsx:
Genes	CASE_1	CASE_2	CASE_3	CASE_4	CASE_5	CASE_6	CTRL_1	CTRL_2	CTRL_3	CTRL_4	CTRL_5	CTRL_6
TSPAN6	9832	25521	23529	0	32	0	10558	20573	6507	0	0	 0
TNMD	3281	 7167	 3435	0	 0	0	  202	  403	 203	0	0	21


SRP021193.colData.tsv:
	diagnosis	group	LibrarySelection
CASE_1	non-ischemic cardiomyopathy	CASE	cDNA


response:
```
# Load necessary packages
library(readxl)
library(tidyverse)

# Read the count table
count_table <- read_excel("Supplementary_Table_1_exp.xlsx")

# Read the annotation data
annotation_data <- read_tsv("SRP021193.colData.tsv")

# Rename the first column of the count table to "Genes"
colnames(count_table)[1] <- "Genes"

# Merge the annotation data and count table
merged_data <- merge(count_table, annotation_data, by.x = "Genes", by.y = "row.names")

# Convert the merged_data dataframe into a tidy format
tidy_data <- merged_data %>% 
  pivot_longer(cols = -Genes, names_to = "Sample", values_to = "Expression")

# Plot boxplots for gene expression values per sample, color-coded based on group variable
ggplot(data = tidy_data, aes(x = Sample, y = Expression, fill = group)) +
  geom_boxplot() +
  xlab("Sample") +
  ylab("Gene Expression")
```

Note: The code assumes that the count table and annotation data file are in the same directory as the R script, or you have provided the full path to the files. Adjust the file paths accordingly if they are located in a different directory. Make sure to replace the sample names (CASE_1, CASE_2, etc.) with the actual sample names present in your count_table dataframe.
