prompt:
I have the following count table for an RNA-seq experiment, rows are genes and columns are samples, "Supplementary_Table_1_exp.xlsx". The first column of the file is my gene names and the rest of the columns are samples. My annotation data is here  "SRP021193.colData.tsv", it contains sample annotations and rows of this matches the columns of count table, first column is the sample ids. Normalize counts, make two PCA plots for samples, one color coded samples by "group" and the other color coded by "LibrarySelection" in my annotation data.

here are first few lines of the file(s).

Supplementary_Table_1_exp.xlsx:
Genes	CASE_1	CASE_2	CASE_3	CASE_4	CASE_5	CASE_6	CTRL_1	CTRL_2	CTRL_3	CTRL_4	CTRL_5	CTRL_6
TSPAN6	9832	25521	23529	0	32	0	10558	20573	6507	0	0	 0
TNMD	3281	 7167	 3435	0	 0	0	  202	  403	 203	0	0	21


SRP021193.colData.tsv:
	diagnosis	group	LibrarySelection
CASE_1	non-ischemic cardiomyopathy	CASE	cDNA


response:
I apologize for the confusion. In R, `read_excel` does not have an argument for `row.names`, which is likely why you are receiving an error. To set the rownames, you can do it separately after reading the table. Below is the corrected code:

```
#Loading required libraries 
library("DESeq2")
library("readxl")
library("ggplot2")

#Reading the data
countData <- read_excel("Supplementary_Table_1_exp.xlsx")
rownames(countData) <- countData$Genes
countData$Genes <- NULL

colData <- read.delim("SRP021193.colData.tsv", row.names = 1)

#Making a DESeqDataSet form the count data and col data
dds <- DESeqDataSetFromMatrix(countData = countData, colData = colData, design = ~ group)

#Running the DESeq normalization
dds <- DESeq(dds)

#Getting the rlog transformed normalized counts
rld <- rlog(dds, blind = TRUE)

#Creating PCA data
pcaData <- plotPCA(rld, intgroup = c("group", "LibrarySelection"), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))

#PCA plot for group
ggplot(pcaData, aes(PC1, PC2, color = group)) +
  geom_point(size = 3) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed() +
  ggtitle("PCA Plot (Color coded by Group)") 

#PCA plot for LibrarySelection
ggplot(pcaData, aes(PC1, PC2, color = LibrarySelection)) +
  geom_point(size = 3) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed() +
  ggtitle("PCA Plot (Color coded by LibrarySelection)") 
```

This code reads the count data excluding the first column as a gene column. Then sets this column to be the row names for the `countData` dataframe. Rest of the script is same as earlier.
