prompt:
I have the following count table for an RNA-seq experiment, rows are genes and columns are samples, "Supplementary_Table_1_exp.xlsx". The first column of the file is my gene names and the rest of the columns are samples. My annotation data is here  "SRP021193.colData.tsv", it contains sample annotations and rows of this matches the columns of count table, first column is the sample ids. Find genes specific for each "group" in my annotation data  using statistical tests. Plot PCA only using "group" specific genes and color code PCA by "group" variable in my annotation data.

here are first few lines of the file(s).

Supplementary_Table_1_exp.xlsx:
Genes	CASE_1	CASE_2	CASE_3	CASE_4	CASE_5	CASE_6	CTRL_1	CTRL_2	CTRL_3	CTRL_4	CTRL_5	CTRL_6
TSPAN6	9832	25521	23529	0	32	0	10558	20573	6507	0	0	 0
TNMD	3281	 7167	 3435	0	 0	0	  202	  403	 203	0	0	21


SRP021193.colData.tsv:
	diagnosis	group	LibrarySelection
CASE_1	non-ischemic cardiomyopathy	CASE	cDNA


response:
```
# Load necessary libraries
library(readxl)
library(dplyr)
library(edgeR)
library(ggplot2)
library(RColorBrewer)

# Read count table and annotation data
count_table <- read_excel("Supplementary_Table_1_exp.xlsx")
annotation_data <- read.delim("SRP021193.colData.tsv", sep = "\t")

# Rename columns in count table
colnames(count_table) <- count_table[1, ]
count_table <- count_table[-1, ]

# Remove unnecessary columns in annotation data
annotation_data <- annotation_data[, c("diagnosis", "group")]

# Perform statistical tests to find group-specific genes using edgeR
rownames(count_table) <- count_table[, 1]
count_table <- count_table[, -1]
y <- DGEList(counts = count_table, group = annotation_data$group)
y <- calcNormFactors(y)
design <- model.matrix(~ group, data = annotation_data)
y <- estimateDisp(y, design)
fit <- glmFit(y, design)
lrt <- glmLRT(fit)
topTags(lrt)

# Select top differentially expressed genes for each group
top_genes <- c()
for (group in unique(annotation_data$group)) {
  top_genes_group <- rownames(lrt)[lrt$table$PValue[lrt$table$group == group] < 0.05]
  top_genes <- c(top_genes, top_genes_group)
}
top_genes <- unique(top_genes)

# Subset count table and annotation data with top genes
count_table_subset <- count_table[top_genes, ]
annotation_data_subset <- annotation_data[rownames(count_table_subset), ]

# Perform PCA using top genes
pca_data <- prcomp(count_table_subset)

# Create a data frame for PCA results
pca_df <- data.frame(PC1 = pca_data$x[, 1], PC2 = pca_data$x[, 2], Group = annotation_data_subset$group)

# Plot PCA with color-coded groups
ggplot(pca_df, aes(x = PC1, y = PC2, color = Group)) +
  geom_point() +
  scale_color_manual(values = brewer.pal(n = length(unique(annotation_data_subset$group)), name = "Set1"))
```
```
