prompt:
I have the following count table for an RNA-seq experiment, rows are genes and columns are samples,"Supplementary_Table_1_exp.xlsx". The first column of the file is my gene names and the rest of the columns are samples. My annotation data is here "SRP021193.colData.tsv", it contains sample annotations and rows of this matches the columns of count table, first column is the sample ids. Merge annotation data and gene expression tables. Plot boxplots for gene expression values per sample and color code boxplots based on group" variable in my annotation data.

response:
```
# Load required libraries
library(readxl)
library(tidyverse)

# Read count table
count_table <- read_excel("Supplementary_Table_1_exp.xlsx")

# Read annotation data
annotation_data <- read_tsv("SRP021193.colData.tsv")

# Set row names of count table as gene names
rownames(count_table) <- count_table$Gene

# Remove the first column of count table as it is duplicated with row names
count_table <- count_table[,-1]

# Merge annotation data and count table based on matching row names
merged_data <- merge(annotation_data, count_table, by.y = "row.names", all = TRUE)

# Plot boxplots for gene expression values per sample, colored based on group variable
merged_data %>%
  pivot_longer(names_to = "Sample", values_to = "Expression", cols = starts_with("Sample")) %>%
  ggplot(aes(x = Group, y = Expression, fill = Group)) +
  geom_boxplot() +
  labs(title = "Boxplot of Gene Expression per Sample",
       x = "Group",
       y = "Gene Expression")
```
Note: Make sure to change the file paths in the `read_excel` and `read_tsv` functions to the correct paths of your files.
